yield_plugin/src/yield_plugin.cpp

```diff
@@ -403,106 +415,243 @@ namespace yield_plugin
     jmt_trajectory.trajectory_points = jmt_trajectory_points;
     return jmt_trajectory;
   }
+
+  void YieldPlugin::external_objects_callback(const carma_perception_msgs::msg::ExternalObjectList::UniquePtr msg)
+  {
+    external_objects_ = msg->objects;
+  }
   
-  carma_planning_msgs::msg::TrajectoryPlan YieldPlugin::update_traj_for_object(const carma_planning_msgs::msg::TrajectoryPlan& original_tp, double initial_velocity) 
+  boost::optional<rclcpp::Time> YieldPlugin::detect_collision_time(const carma_planning_msgs::msg::TrajectoryPlan& trajectory1, const std::vector<carma_perception_msgs::msg::PredictedState>& trajectory2, double collision_radius) 
   {
-        
+    // Iterate through each pair of consecutive points in the trajectories
+
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Starting a new collision detection, trajectory size: " << trajectory1.trajectory_points.size() << ". prediction size: " << trajectory2.size());
+
+    // Iterate through the object to check if it's on the route
+    bool on_route = false;
+    int on_route_idx = 0;
+
+    for (int j = 0; j < trajectory2.size(); ++j) 
+    {
+      lanelet::BasicPoint2d curr_point;
+      curr_point.x() = trajectory2[j].predicted_position.position.x;
+      curr_point.y() = trajectory2[j].predicted_position.position.y;
+
+      auto corresponding_lanelets = wm_->getLaneletsFromPoint(curr_point, 4); // get 4 lanelets
+
+      for (auto llt: corresponding_lanelets)
+      {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Checking llt: " << llt.id());
+
+        if (route_llt_ids_.find(llt.id()) != route_llt_ids_.end())
+        {
+          on_route = true;
+          on_route_idx = j;
+          break;
+        }
+      }
+      if (on_route)
+        break;
+    }
+
+    if (!on_route)
+    {
+      RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Predicted states are not on the route! ignoring");
+      return boost::none;
+    }
+    
+    double smallest_dist = std::numeric_limits<double>::max();
+    for (int i = 0; i < trajectory1.trajectory_points.size() - 1; ++i) 
+    {
+      auto p1a = trajectory1.trajectory_points[i];
+      auto p1b = trajectory1.trajectory_points[i + 1];
+      for (int j = on_route_idx; j < trajectory2.size() - 1; ++j) 
+      {
+        auto p2a = trajectory2[j];
+        auto p2b = trajectory2[j + 1];
```

Comment:
We should be accessing elements using `.at()` instead of `operator[]` unless performance is critical. I think the bounds checking outweighs the performance hit here.